types/validator_set.go

```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
 	}
 }
 
+// Checks changes against duplicates, splits the changes in updates and removals, sorts them by address
+func processChanges(ochanges []*Validator) (updates, removals []*Validator, err error) {
+	// Make a deep copy of the changes and sort by address
+	changes := validatorListCopy(ochanges)
+	sort.Sort(ValidatorsByAddress(changes))
+
+	removals = make([]*Validator, 0)
+	updates = make([]*Validator, 0)
+	prevAddr := []byte(nil)
+
+	// Scan changes by address and append valid validators to updates or removal
+	for _, valUpdate := range changes {
+		if bytes.Equal(valUpdate.Address, prevAddr) {
+			err := fmt.Errorf("duplicate entry %v in changes list %v", valUpdate, changes)
+			return nil, nil, err
+		}
+		if valUpdate.VotingPower < 0 {
+			err := fmt.Errorf("voting power can't be negative %v", valUpdate)
+			return nil, nil, err
+		}
+		if valUpdate.VotingPower == 0 {
+			// Add valUpdate to removals
+			removals = append(removals, valUpdate)
+		} else {
+			// Add valUpdate to updates
+			updates = append(updates, valUpdate)
+		}
+		prevAddr = valUpdate.Address
+	}
+	return updates, removals, nil
+}
+
+// Verifies a list of updates against a validator set, making sure the allowed
+// total voting power would not exceed if these updates would be applied to the set.
+// Computes the proposer priority for the validators not present in the set.
+// Returns error if first step fails.
+// updates parameter must be a list of unique validators to be added or updated.
```

Comment:
Maybe add a comment that explains that this modifies the priorities of the passed `updates` but does not change `vals`?